hydra-main/src/main/java/com/addthis/hydra/query/web/HttpStaticFileHandler.java

```diff
@@ -179,35 +179,32 @@ public void channelRead0(
 
         log.trace("sending {}", file);
 
-        FileChannel fileChannel;
+        FileChannel fileChannel = null;
         try {
             fileChannel = FileChannel.open(file, StandardOpenOption.READ);
-        } catch (IOException fnfe) {
-            sendError(ctx, NOT_FOUND);
-            return;
-        }
-        long fileLength = fileChannel.size();
+            long fileLength = fileChannel.size();
 
-        HttpResponse response = new DefaultHttpResponse(HTTP_1_1, OK);
-        setContentLength(response, fileLength);
-        setContentTypeHeader(response, file);
-        try {
+            HttpResponse response = new DefaultHttpResponse(HTTP_1_1, OK);
+            setContentLength(response, fileLength);
+            setContentTypeHeader(response, file);
             setDateAndCacheHeaders(response, file);
+            if (isKeepAlive(request)) {
+                response.headers().set(CONNECTION, HttpHeaders.Values.KEEP_ALIVE);
+            }
+
+            // Write the initial line and the header.
+            ctx.write(response);
+
+            // Write the content.
+            ctx.write(new DefaultFileRegion(fileChannel, 0, fileLength));
```

Comment:
Netty can be kind of a confusing beast, so I'm not surprised your tools got tripped up here. There's a couple issues, some rather significant:

1) the fileChannel should not be closed right after `ctx.write` is called. That write is asynchronous, and it probably won't even try to write at all until that writeAndFlush is called later. In this case, DefaultFileRegion has some error recovery logic where it will construct a new RAF to get a channel from as long as it still has leases left, but that shouldn't be relied on and it breaks the lease paradigm as seen in

2) As soon as we put the fileChannel into that wrapper and give it to `ctx.write`, it is no longer our business to close it. It will be closed in `DefaultFileRegion`'s deallocate method when the latter runs out of leases (pretty much meaning in this case, when it has been fully written to the socket).

3) As soon as `ctx.write(response)` was called, then queuing up the http response header et al, we should strongly prefer to avoid any code paths that could lead to eg. `sendError(ctx, NOT_FOUND)` being called. Most http clients become extremely upset when you send them conflicting http response codes.

4) The separate catch blocks for the two sources of `IOException`s was not my original design (much of this class was appropriated from netty), but I think I would have gone with something not too dissimilar anyway. Of all the exception classes, `IOException` is the most common, and I would even say overused even if it isn't by its own choice (being checked and all, it gets forwarded everywhere). Although neither version provides as much information as it could, the base version from netty at least makes a compelling case that the author was sure the file vanished (for a moment, or for however long), and there's only two lines to double check that assertion. If I was returning to this class after a long time tracking down such an error, I highly doubt I would immediately suspect the now innocuously placed `setDateAndCacheHeaders` from being any more risky than its neighbors.